# error

Una [Herramienta Sumor Cloud](https://sumor.cloud).  
[Más Documentación](https://sumor.cloud/error)
Biblioteca de manejo de errores con soporte para múltiples idiomas, código de error predefinido y paso de datos.

[![CI](https://github.com/sumor-cloud/error/actions/workflows/ci.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ci.yml)
[![Test](https://github.com/sumor-cloud/error/actions/workflows/ut.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ut.yml)
[![Cobertura](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml)
[![Revisión](https://github.com/sumor-cloud/error/actions/workflows/audit.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/audit.yml)

## Instalación

```bash
npm i @sumor/error --save
```

## Prerrequisitos

### Versión de Node.JS

Se requiere Node.JS versión 18.x o superior

### Requiere módulo ES de Node.JS

Dado que este paquete está escrito en módulo ES,
por favor cambie el siguiente código en su archivo `package.json`:

```json
{
  "type": "module"
}
```

## Uso

### Definir Error

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Usuario no encontrado',
    USER_EXISTED: 'Usuario {name} existía'
  }
})

throw new MyError('USER_NOT_FOUND')
// output: Error: Usuario no encontrado

throw new MyError('USER_EXISTED', { name: 'Alice' })
// output: Error: Usuario Alice existía
```

### Multi-idioma

```js
import defineError from '@sumor/error'

const MyError = defineError({
  language: 'es', // idioma predeterminado
  code: {
    USER_NOT_FOUND: 'Usuario no encontrado',
    USER_EXISTED: 'Usuario {name} existía'
  },
  i18n: {
    en: {
      USER_NOT_FOUND: 'User not found',
      USER_EXISTED: 'User {name} existed'
    }
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
error.language = 'es' // cambiar idioma del error
console.log(error)
// output: Error: Usuario Alice existía

error.language = 'en' // cambiar idioma del error
console.log(error)
// output: Error: User Alice existed
```

### Convertir Error a JSON

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Usuario no encontrado',
    USER_EXISTED: 'Usuario {name} existía'
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
console.log(error.json())
// output: {"code":"USER_EXISTED","message":"Usuario Alice existía"}
```

### Error Subyacente

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Fallo en la verificación del campo',
    FIELD_CANNOT_EMPTY: 'El campo {name} no puede estar vacío',
    FIELD_TOO_LONG: 'El campo {name} es demasiado largo'
  },
  i18n: {
    en: {
      FIELD_VERIFY_FAILED: 'Field verify failed',
      FIELD_CANNOT_EMPTY: 'Field {name} cannot be empty',
      FIELD_TOO_LONG: 'Field {name} is too long'
    }
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'nombre de usuario' }),
  new MyError('FIELD_TOO_LONG', { name: 'contraseña' })
])

console.log(error.json())
/* 
output: 
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Fallo en la verificación del campo",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"El campo nombre de usuario no puede estar vacío"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"El campo contraseña es demasiado largo"
    }
  ]
}
*/

error.language = 'en'
console.log(error.json())
/*
output:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Field verify failed",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"Field username cannot be empty"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"Field password is too long"
    }
  ]
}
*/
```

### Combinar Error Estándar

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Fallo en la verificación del campo',
    FIELD_CANNOT_EMPTY: 'El campo {name} no puede estar vacío',
    FIELD_TOO_LONG: 'El campo {name} es demasiado largo'
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'nombre de usuario' }),
  new MyError('FIELD_TOO_LONG', { name: 'contraseña' }),
  new Error('Error Desconocido')
])

console.log(error.json())
/*
output:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Fallo en la verificación del campo",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"El campo nombre de usuario no puede estar vacío"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"El campo contraseña es demasiado largo"
    },{
      "code":"UNKNOWN_ERROR",
      "message":"Error Desconocido"
    }
  ]
}
 */
```
